File: apps/server/src/adapter/driven/auth/jwt.ts
import jwt from "jsonwebtoken";
import type { TokenService, TokenClaims } from "../../../domain/port/index.js";

export interface JwtConfig {
  accessSecret: string;
  refreshSecret: string;
  accessExpiresIn: string;  // e.g. "15m"
  refreshExpiresIn: string; // e.g. "7d"
}

interface JwtPayload {
  sub: string;
  email: string;
  role: string;
  permissions: string[];
  type: "access" | "refresh";
}

export class JwtTokenService implements TokenService {
  private readonly config: JwtConfig;

  constructor(config: JwtConfig) {
    this.config = config;
  }

  generateAccessToken(claims: TokenClaims | { id: string; email: string; role: string; permissions?: string[] }): string {
    const userId = "userId" in claims ? claims.userId : claims.id;
    const permissions = "permissions" in claims ? (claims.permissions ?? []) : [];
    const payload: JwtPayload = {
      sub: userId,
      email: claims.email,
      role: claims.role,
      permissions,
      type: "access",
    };
    return jwt.sign(payload, this.config.accessSecret, {
      expiresIn: this.config.accessExpiresIn as any,
    });
  }

  generateRefreshToken(claims: TokenClaims | { id: string; email: string; role: string; permissions?: string[] }): string {
    const userId = "userId" in claims ? claims.userId : claims.id;
    const payload: Omit<JwtPayload, "permissions"> & { type: "refresh" } = {
      sub: userId,
      email: claims.email,
      role: claims.role,
      type: "refresh",
    };
    return jwt.sign(payload, this.config.refreshSecret, {
      expiresIn: this.config.refreshExpiresIn as any,
    });
  }

  verifyRefreshToken(token: string): { userId: string } {
    const decoded = this.validateRefreshToken(token);
    return { userId: decoded.userId };
  }

  validateAccessToken(token: string): TokenClaims & { permissions: string[] } {
    const decoded = jwt.verify(token, this.config.accessSecret) as JwtPayload;
    if (decoded.type !== "access") {
      throw new Error("Invalid token type: expected access token");
    }
    return {
      userId: decoded.sub,
      email: decoded.email,
      role: decoded.role,
      permissions: decoded.permissions ?? [],
    };
  }

  validateRefreshToken(token: string): TokenClaims {
    const decoded = jwt.verify(token, this.config.refreshSecret) as JwtPayload;
    if (decoded.type !== "refresh") {
      throw new Error("Invalid token type: expected refresh token");
    }
    return {
      userId: decoded.sub,
      email: decoded.email,
      role: decoded.role,
    };
  }
}

File: apps/api/internal/adapter/driven/auth/jwt.go
package auth

import (
	"errors"
	"time"

	"github.com/golang-jwt/jwt/v5"
	"github.com/neurodynecorp/api/internal/domain/port"
)

type JWTService struct {
	secret          []byte
	expirationHours int
	refreshHours    int
}

func NewJWTService(secret string, expirationHours, refreshHours int) *JWTService {
	return &JWTService{
		secret:          []byte(secret),
		expirationHours: expirationHours,
		refreshHours:    refreshHours,
	}
}

type Claims struct {
	UserID string `json:"user_id"`
	Role   string `json:"role,omitempty"`
	Type   string `json:"type"`
	jwt.RegisteredClaims
}

func (s *JWTService) GenerateAccessToken(userID string, role string) (string, error) {
	claims := Claims{
		UserID: userID,
		Role:   role,
		Type:   "access",
		RegisteredClaims: jwt.RegisteredClaims{
			ExpiresAt: jwt.NewNumericDate(time.Now().Add(time.Duration(s.expirationHours) * time.Hour)),
			IssuedAt:  jwt.NewNumericDate(time.Now()),
			Issuer:    "neurodynecorp",
		},
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	return token.SignedString(s.secret)
}

func (s *JWTService) GenerateRefreshToken(userID string) (string, error) {
	claims := Claims{
		UserID: userID,
		Type:   "refresh",
		RegisteredClaims: jwt.RegisteredClaims{
			ExpiresAt: jwt.NewNumericDate(time.Now().Add(time.Duration(s.refreshHours) * time.Hour)),
			IssuedAt:  jwt.NewNumericDate(time.Now()),
			Issuer:    "neurodynecorp",
		},
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	return token.SignedString(s.secret)
}

func (s *JWTService) ValidateAccessToken(tokenStr string) (*port.TokenClaims, error) {
	return s.validateToken(tokenStr, "access")
}

func (s *JWTService) ValidateRefreshToken(tokenStr string) (*port.TokenClaims, error) {
	return s.validateToken(tokenStr, "refresh")
}

func (s *JWTService) validateToken(tokenStr string, expectedType string) (*port.TokenClaims, error) {
	token, err := jwt.ParseWithClaims(tokenStr, &Claims{}, func(token *jwt.Token) (interface{}, error) {
		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, errors.New("unexpected signing method")
		}
		return s.secret, nil
	})
	if err != nil {
		return nil, err
	}

	claims, ok := token.Claims.(*Claims)
	if !ok || !token.Valid {
		return nil, errors.New("invalid token")
	}

	if claims.Type != expectedType {
		return nil, errors.New("invalid token type")
	}

	return &port.TokenClaims{
		UserID: claims.UserID,
		Role:   claims.Role,
	}, nil
}

var _ port.TokenService = (*JWTService)(nil)
